src/StepAdaptationController/src/StepAdaptationController.cpp

```diff
@@ -487,9 +517,364 @@ bool StepAdaptationController::getAdaptatedFootTrajectory(const footTrajectoryGe
+            rightArmRollError= abs(qDesired(var)-qActual(var));
+        }
+    }
+
+    if (leftArmPitchError>getRollPitchErrorThreshold()(1) )
+    {
+        m_armPitchError=leftArmPitchError;
+        m_isPitchActive=1;
+    }
+    else if( rightArmPitchError>getRollPitchErrorThreshold()(1))
+    {
+        m_armPitchError=rightArmPitchError;
+        m_isPitchActive=1;
+    }
+    else
+    {
+        m_armPitchError=0;
+    }
+
+    if (leftArmRollError>getRollPitchErrorThreshold()(0) )
+    {
+        if (leftInContact.front())
+        {
+            m_armRollError=1*leftArmRollError;
+        }
+        if (rightInContact.front())
+        {
+            m_armRollError=-1*leftArmRollError;
+        }
+        m_isRollActive=1;
+    }
+    else if( rightArmRollError>getRollPitchErrorThreshold()(0) )
+    {
+        if (rightInContact.front())
+        {
+            m_armRollError=-1*rightArmRollError;
+        }
+        if (leftInContact.front())
+        {
+            m_armRollError=+1*rightArmRollError;
+        }
+        m_isRollActive=1;
+    }
+    else
+    {
+        m_armRollError=0;
+    }
+
+    return true;
+}
+
+const double& StepAdaptationController::getArmRollError() const
+{
+    return m_armRollError;
+}
+
+const double& StepAdaptationController::getArmPitchError() const
+{
+    return m_armPitchError;
+}
+
+bool StepAdaptationController::UpdateDCMEstimator(const iDynTree::Vector2& CoM2DPosition,const iDynTree::Vector2& CoMVelocity,const iDynTree::Vector2& measuredZMP,const double& CoMHeight)
+{
+    iDynTree::Rotation imuRotation;
+    iDynTree::Vector3 imuRPY;
+    iDynTree::Rotation StanceFootOrientation;
+    StanceFootOrientation=iDynTree::Rotation::Identity();
+
+    if (m_isRollActive)
+    {
+    m_armRollError= m_armRollError+0.1;
+    }
+    if (m_isPitchActive)
+    {
+    m_armPitchError= m_armPitchError+0.1;
+    }
+
+    StanceFootOrientation=iDynTree::Rotation::RPY (m_armRollError,m_armPitchError,0); 	//=m_FKSolver->getRootLinkToWorldTransform().getRotation().asRPY();
+
+    iDynTree::Vector3 ZMP3d;
+    ZMP3d(0)=measuredZMP(0);
+    ZMP3d(1)=measuredZMP(1);
+    ZMP3d(2)=0;
+
+    iDynTree::Vector3 CoM3d;
+    //iDynTree::Vector3 ZMP3d;
+    CoM3d(0)=CoM2DPosition(0);
+    CoM3d(1)=CoM2DPosition(1);
+    CoM3d(2)=CoMHeight;
+
+    iDynTree::Vector3 CoMVelocity3d;
+    //iDynTree::Vector3 ZMP3d;
+    CoMVelocity3d(0)=CoMVelocity(0);
+    CoMVelocity3d(1)=CoMVelocity(1);
+    CoMVelocity3d(2)=0;
+
+    if(!m_DCMEstimator->update(StanceFootOrientation,ZMP3d,CoM3d,CoMVelocity3d))
+    {
+        yError() << "[StepAdaptationController::UpdateDCMEstimator] Unable to to recieve DCM from pendulumEstimator";
+        return false;
+    }
+
+    return true;
+}
+
+bool StepAdaptationController::runStepAdaptation(const runningStepAdapterInput& input, runStepAdapterOutput& output)
+{
+    if (!input.leftInContact.front() || !input.rightInContact.front())
+    {
+        output.indexPush=output.indexPush+1;
```

Comment:
Why ``indexPush`` is needed?

Reply:
It's described here:
https://github.com/MiladShafiee/walking-controllers/blob/93b8534cbee263efbf26434ba9372658f4b946e9/src/WalkingModule/include/WalkingControllers/WalkingModule/Module.h#L81

Reply:
I did not notice it! Thanks.